src/java.base/share/classes/java/lang/reflect/Field.java

```diff
@@ -829,6 +882,9 @@ public void set(Object obj, Object value)
             checkAccess(caller, obj);
             getFieldAccessor().set(obj, value);
         } else {
+            if (isFinalInstanceInNormalClass() && obj != null) {
+                checkAllowedToSetFinal(Reflection.getCallerClass());
+            }
```

Comment:
There are multiple duplicate codes here, or put them all in one method, like this

```java
    @ForceInline
    private void checkAllowedToSetFinalIfFinalInstanceInNormalClass(Object obj) throws IllegalAccessException {
        if (Modifier.isFinal(modifiers)
                && !Modifier.isStatic(modifiers)
                && !clazz.isRecord()
                && !clazz.isHidden()
                && obj != null) {
            checkAllowedToSetFinal(Reflection.getCallerClass(), false);
        }
    }
```